internal/execute/solutionbuilder.go

```diff
@@ -0,0 +1,690 @@
+}
+
+func (s *solutionBuilder) Clean() CommandLineResult {
+	s.setup()
+	var buildResult solutionBuilderResult
+	if len(s.orderGenerator.errors) == 0 {
+		wg := core.NewWorkGroup(s.opts.command.CompilerOptions.SingleThreaded.IsTrue())
+		s.cleanProjects(wg, &buildResult)
+		wg.RunAndWait()
+		buildResult.statistics.projects = len(s.orderGenerator.Order())
+	} else {
+		s.buildResultOfCircularOrder(&buildResult)
+	}
+	buildResult.report(s)
+	return buildResult.result
+}
+
+func (s *solutionBuilder) buildResultOfCircularOrder(buildResult *solutionBuilderResult) {
+	buildResult.result.Status = ExitStatusProjectReferenceCycle_OutputsSkipped
+	reportDiagnostic := s.createDiagnosticReporter(nil)
+	for _, err := range s.orderGenerator.errors {
+		reportDiagnostic(err)
+	}
+	buildResult.errors = s.orderGenerator.errors
+}
+
+func (s *solutionBuilder) relativeFileName(fileName string) string {
+	return tspath.ConvertToRelativePath(fileName, s.comparePathsOptions)
+}
+
+func (s *solutionBuilder) toPath(fileName string) tspath.Path {
+	return tspath.ToPath(fileName, s.comparePathsOptions.CurrentDirectory, s.comparePathsOptions.UseCaseSensitiveFileNames)
+}
+
+func (s *solutionBuilder) setup() {
+	s.host = &solutionBuilderHost{
+		builder: s,
+		host:    compiler.NewCachedFSCompilerHost(s.opts.sys.GetCurrentDirectory(), s.opts.sys.FS(), s.opts.sys.DefaultLibraryPath(), nil, nil),
+	}
+	s.orderGenerator = NewBuildOrderGenerator(s.opts.command, s.host, s.opts.command.CompilerOptions.SingleThreaded.IsTrue())
+}
+
+func (s *solutionBuilder) buildProjects(wg core.WorkGroup, projects []string, buildResult *solutionBuilderResult, seen *collections.SyncSet[tspath.Path]) {
+	for _, project := range projects {
+		s.startProjectBuild(wg, project, buildResult, seen)
+	}
+}
+
+func (s *solutionBuilder) startProjectBuild(wg core.WorkGroup, config string, buildResult *solutionBuilderResult, seen *collections.SyncSet[tspath.Path]) {
+	path := s.toPath(config)
+	if !seen.AddIfAbsent(path) {
+		return // Already seen this project
+	}
+	wg.Queue(func() {
+		task, ok := s.orderGenerator.tasks.Load(path)
+		if !ok {
+			panic("No build task found for " + config)
+		}
+
+		// Queue the upstream tasks
+		for _, upstream := range task.upStream {
+			if upstream.status != nil {
+				s.startProjectBuild(wg, upstream.config, buildResult, seen)
+			}
+		}
+
+		// Wait on upstream tasks to complete
+		upStreamStatus := make([]*upToDateStatus, len(task.upStream))
+		for i, upstream := range task.upStream {
+			if upstream.status != nil {
+				upStreamStatus[i] = <-upstream.status
+			}
+		}
+
+		status, taskReporter := s.buildProject(config, path, task, upStreamStatus)
+		for _, downstream := range task.downStream {
+			downstream.status <- status
+		}
+
+		// Wait for previous build task to complete reporting status, errors etc
+		if task.previousTaskReporter != nil {
+			<-task.previousTaskReporter
+		}
+		taskReporter.report(s, path, buildResult)
+		task.reporter <- taskReporter
```

Comment:
Would it be simpler to just make the reporter have a mutex, rather than trying to hand things around?